src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
+        }
+
+        private static T[] LazyToArray<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(!(source is ICollection<T>), $"We should have gone down the optimized route if {nameof(source)} was a regular collection.");
+
+            const int InitialCapacity = 4;
+            const int BufferListThreshold = 32;
+
+            using (var en = source.GetEnumerator())
+            {
+                if (!en.MoveNext())
+                    return Array.Empty<T>();
+
+                T[] buffer = null;
+                int index = 0; // index into the buffer we are reading into
+
+                do
+                {
+                    if (buffer == null)
+                    {
+                        // First iteration: allocate an array with the initial capacity
+                        buffer = new T[InitialCapacity];
+                    }
+                    else
+                    {
+                        // Resize from a previous iteration
+                        T[] last = buffer;
+                        int newLength = last.Length * 2;
+                        buffer = new T[newLength];
+
+                        // We are using a for-loop instead of Array.Resize under the assumption
+                        // we are copying < 32 elements.
+                        Debug.Assert(last.Length < BufferListThreshold);
+                        for (int i = 0; i < last.Length; i++) buffer[i] = last[i];
+                    }
+
+                    do
+                    {
+                        buffer[index++] = en.Current;
+
+                        if (index == buffer.Length)
+                        {
+                            // We used up the last slot
+                            break;
+                        }
+                    }
+                    while (en.MoveNext());
+
+                    if (index < buffer.Length)
+                    {
+                        // Since MoveNext returned false, this means that we finished enumerating
+                        // before we reached the end of the buffer.
+                        int finalLength = index;
+                        Debug.Assert(finalLength < buffer.Length); // case where they're == is taken care of below
+                        
+                        // Since we only arrive here for enumerables of length < 32,
+                        // copy using a manual for-loop rather than Array.Resize, which
+                        // does not perform as well for smaller arrays.
+                        var result = new T[finalLength];
+                        for (int i = 0; i < result.Length; i++) result[i] = buffer[i];
+                        return result;
+                    }
+
+                    // We called Current last, but the loop expects MoveNext to be called
+                    // before it is entered. So we need to call it again here.
+                    if (!en.MoveNext())
```

Comment:
IEnumerator.MoveNext is spec'd to return false on subsequent calls after it's been called, but I'm concerned that not every IEnumerator implementation follows that, and this is introducing an additional MoveNext call after MoveNext may have returned false in certain cases.  Let's avoid that.